### ISSUE TITLE
Timeout when searching for number fields by abstract Galois group

### ISSUE THREAD

**Issue by AndrewVSutherland** [LMFDB/lmfdb#3215]
If you go to http://www.lmfdb.org/NumberField/ and search for all number fields with a particular Galois group specified as an abstract group (e.g. D4 or A5) without specifying the degree the search can be very slow and will often simply time out.  If one also specifies the degree or uses a transitive group label the search is much faster, but most first time users (and even experienced users) will not think to do this.

The reason for the problem is that the nf_fields table only knows Galois groups as transitive groups, and the same abstract group can be realized as many different transitive groups (e.g. A5 could be any of 5T4, 6T17, 10T7, 12T33, 15T5, 20T15, 30T9), and searching on a union of key values is slow (mostly because of the need to get the sort order right).

One way to solve the problem would be to add a group label column to nf_fields (in addition to the galt column we already have).  This column would hold just a single string, which is either a short name (e.g. Cn, Dn, Sn, An, GL(2,q), SL(2,q) PSL(2,q), Q8, ...) or None if the isomorphism class of the Galois group does not have a short name that we support.  In cases where the same group has one more than one short name (e.g. D3=S3, A5=PSL(2,5)) we should pick one to store and treat the other(s) as an alias that we automatically convert to our chosen short name -- in fact we already do this (entering PSL(2,5) returns A5 fields).

Another way to solve the problem would be to figure out a way to speed up the union query (possibly by analyzing the postgres query plan and tweaking indexes or the query that is used).

Note: if you are reading this shortly after I post this issue and try the queries I suggest, they are likely to run very quickly due to caching (I have been trying out lots of queries while writing this issue).  But if you set an alarm for a random time in the next 24-48 hours and search for all D4 fields at that time, I predict the query will take at least 10-15 seconds and may well time out (as it did for me 15 minutes ago).

**Comment by edgarcosta** [comment]
At the moment queries like the ones done through the web interface rely solely on the index to help them sort:
```
EXPLAIN ANALYZE SELECT "coeffs", "degree", "rd", "iso_number", "num_ram", "label", "ramps_array", "coeffs_array", "cm", "ramps", "regulator", "class_number", "galt", "disc_rad", "r2", "disc_abs", "used_grh", "disc_sign", "class_group" FROM "nf_fields" WHERE ("galt" = 4 AND "degree" = 8 OR "galt" = 3 AND "degree" = 4) ORDER BY "degree", "disc_abs", "disc_sign", "iso_number" LIMIT 1000;
                                                                                      QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.56..9124.14 rows=1000 width=356) (actual time=1150.821..1152.854 rows=1000 loops=1)
   ->  Index Scan using nf_fields_degree_disc_abs_disc_sign_iso_number on nf_fields  (cost=0.56..3551929.80 rows=389313 width=356) (actual time=1150.819..1152.761 rows=1000 loops=1)
         Filter: (((galt = 4) AND (degree = 8)) OR ((galt = 3) AND (degree = 4)))
         Rows Removed by Filter: 2239032
 Planning time: 0.364 ms
 Execution time: 1152.960 ms
```

and even after adding an index that includes every sorting column and `galt` 
```
CREATE INDEX nf_fields_degree_disc_abs_disc_sign_iso_number_galt_tmp on nf_fields using BTREE ("degree", "disc_abs", "disc_sign", "iso_number", "galt");
```

the behavior doesn't change.

The problem with this specific query is that about 10% of the number fields satisfy the constraint 
```"galt" = 3 AND "degree" = 4```
However, these are not uniformly distributed, and postgres doesn't know that.
Indeed, @roed314's theory is that postgres just decides to do a sequential ordered scan as it hopes to find the desired first 1000 results in the first 10000 rows or so, but obviously before seeing anything interesting must loop through the quadratic and cubic number fields...

Another annoying detail is that postgres doesn't realize that the degree in the constraint is also used for sorting. For example, if one drops the degree 8 case, we get much more sane use of indexes:
```
EXPLAIN ANALYZE SELECT "coeffs", "degree", "rd", "iso_number", "num_ram", "label", "ramps_array", "coeffs_array", "cm", "ramps", "regulator", "class_number", "galt", "disc_rad", "r2", "disc_abs", "used_grh", "disc_sign", "class_group" FROM "nf_fields" WHERE ("galt" = 3 AND "degree" = 4) ORDER BY "degree", "disc_abs", "disc_sign", "iso_number" LIMIT 1000;
                                                                                       QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.56..2248.58 rows=1000 width=356) (actual time=0.053..1.716 rows=1000 loops=1)
   ->  Index Scan using nf_fields_degree_disc_abs_disc_sign_iso_number_galt_tmp on nf_fields  (cost=0.56..807064.84 rows=359011 width=356) (actual time=0.052..1.633 rows=1000 loops=1)
         Index Cond: ((degree = 4) AND (galt = 3))
 Planning time: 0.381 ms
 Execution time: 1.798 ms
```
 
I expect the same behavior if one replaces `galt` by @AndrewVSutherland's proposed column.

PS: if we would only sort on `id`, we observe the same behaviour, i.e., a query like 
```"galt" = 4 AND "degree" = 8 OR "galt" = 3 AND "degree" = 4```
will only use the index on `id`
while
```"galt" = 3 AND "degree" = 4```
uses the index on `galt`, `degree`, and `id`

**Comment by roed314** [comment]
See #3244, which takes the second approach that Drew suggested.

**Comment by AndrewVSutherland** [comment]
Fixed by #3244, and now in production.  Searching for all D4 number fields is now nearly instantaneous.

 Thanks David!
